packages/x-data-grid-pro/src/hooks/features/dataSource/utils.ts

```diff
@@ -112,3 +126,78 @@ export class NestedDataManager {
 
   public getActiveRequestsCount = () => this.pendingRequests.size + this.queuedRequests.size;
 }
+
+/**
+ * Provides better cache hit rate by splitting the data into smaller chunks
+ * Splits the data into smaller chunks to be stored in the cache
+ * Merges multiple cache entries into a single response
+ */
+export class CacheChunkManager {
+  private chunkSize: number;
+
+  constructor(config: GridDataSourceCacheChunkManagerConfig) {
+    this.chunkSize = config.chunkSize;
+  }
+
+  public getCacheKeys = (key: GridGetRowsParams) => {
+    if (this.chunkSize < 1 || typeof key.start !== 'number') {
+      return [key];
+    }
+
+    // split the range into chunks
+    const chunkedKeys: GridGetRowsParams[] = [];
+    for (let i = key.start; i < key.end; i += this.chunkSize) {
+      const end = Math.min(i + this.chunkSize - 1, key.end);
+      chunkedKeys.push({ ...key, start: i, end });
+    }
+
+    return chunkedKeys;
+  };
+
+  public splitResponse = (key: GridGetRowsParams, response: GridGetRowsResponse) => {
+    const cacheKeys = this.getCacheKeys(key);
+    const responses = new Map<GridGetRowsParams, GridGetRowsResponse>();
+    cacheKeys.forEach((chunkKey) => {
+      const isLastChunk = chunkKey.end === key.end;
+      const responseSlice: GridGetRowsResponse = {
+        ...response,
+        pageInfo: {
+          ...response.pageInfo,
+          // If the original response had page info, update that information for all but last chunk and keep the original value for the last chunk
+          hasNextPage:
+            (response.pageInfo?.hasNextPage !== undefined && !isLastChunk) ||
```

Comment:
I went with the logic that if you didn't set this you don't need/use it, so I didn't try to append our own data

Reply:
I imagine it might cause issues in some use cases. 

Consider a use-case.

```tsx
pageSizeOptions = [5, 10, 15, 20];
pageSize = 20
// total rows on server: 20

// Call 1 response:
{ rows: [...twentyRows], pageInfo: { hasNextPage: false } }

// The cache manager splits this response into 4 chunks of size 5 each
// all having the same pageInfo as the original response

// The user updates `pageSize`
newPageSize = 5

// The data will be acquired from the cache (first chunk)
// Since the `pageInfo.hasNextPage` is replicated in all chunks as false
// the Data Grid assumes there's no more page and rest of 15 records are never shown
```

Does that make sense?

Reply:
Actually, this was my intention
This is a bug
I thought that your question was should we set it to true even if the original is `undefined` (which we could because obviously there are more pages)
I will update this.
To summarize, logic is as following
you don't set anything, we keep it like that for all chunks
you set anything, we set true for all but the last chunk and the last chunk gets the original value

cursor is done in the same way, so I am not sure why this one had a different condition that had an issue in it :)

thanks